src/battle_util.c

```diff
@@ -4335,6 +4295,30 @@ static u8 ForewarnChooseMove(u32 battler)
     Free(data);
 }
 
+bool8 ChangeTypeBasedOnTerrain(u8 battlerId)
+{
+    u8 battlerType;
+    u16 terrainFlags = VarGet(VAR_TERRAIN) & STATUS_FIELD_TERRAIN_ANY;
```

Comment:
I'm confused what this is doing. VAR_TERRAIN is only used for setting terrain at the start of battle. We only need to check `gFieldStatuses & <flag>`, so lines 4301-4305 look like they can be removed

Reply:
It's for checking the permanent mode of the field statuses. It was straight up copy-pasted and adapted from the checks inside the `case ABILITYEFFECT_SWITCH_IN_TERRAIN` of `AbilityBattleEffects` :eyes:

Reply:
But why does permanence matter? If we're calling this function the terrain has changed to something so we don't care if it's permanent or not

Reply:
I wasn't sure if it was needed or not tbh. It's just something I took into consideration, like "Oh, I should make the function compatible with that, right?" or something. I'll go check if the ability still works for overworld terrains after removing those lines.

Reply:
Scratch that, I'll need to do further testing and looking into. I was sure everything was working correctly, but it's not. The ability doesn't trigger for terrain-inducing overworld weather conditions. I'll look into it either later or tomorrow, I'm just too tired right now.

Reply:
should be able to just tack on `tryactivatechangeabilities` loop at the end of `BattleScript_TerrainSeedLoop` I would think

Reply:
Do you mean a `call BattleScript_ActivateSwitchInAbilities` in `BattleScript_TerrainSeedLoop_NextBattler` before the `return`? Will do :eyes:

I also noticed that I forgot to include to activate the `ABILITYEFFECT_ON_WEATHER` on `Cmd_switchineffects` so I did that as well, and added activations for them at `TryDoEventsBeforeFirstTurn` too, in the same loop used by the activation of `ABILITYEFFECT_ON_SWITCHIN` :eyes: